File: ex475.cpp
class Solution {
public:
    int findRadius(vector<int>& houses, vector<int>& heaters) {
        int m = houses.size(), n = heaters.size();
        if (m == 0)    return 0;
        if (n == 0)    return INT_MAX;
        
        sort(houses.begin(), houses.end());
        sort(heaters.begin(), heaters.end());
    
        int p = 0, MAX = 0;
        for (int i = 0; i < m; i++) {
            while (p + 1 < n && abs(houses[i] - heaters[p]) >= abs(houses[i] - heaters[p + 1]))
                p++;
            MAX = max(MAX, abs(houses[i] - heaters[p]));
        }
        
        return MAX;
    }
};

File: code/ex475.java
// Binary Search

class Solution {
    public int findRadius(int[] houses, int[] heaters) {
        Arrays.sort(houses);
        Arrays.sort(heaters);
        
        int res = 0;
        for (int house : houses) {
            int dis = distToClosestHeater(heaters, house);
            // System.out.println(dis);
            res = Math.max(res, dis);
        }
        
        return res;
    }
    
    private int distToClosestHeater(int[] heaters, int target) {
        int lo = 0, hi = heaters.length - 1;
        
        int min = Integer.MAX_VALUE;
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            min = Math.min(min, Math.abs(heaters[mid] - target));
            
            if (heaters[mid] < target)
                lo = mid + 1;
            else
                hi = mid - 1;
        }
        
        return min;
    }
}